casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala

```diff
@@ -91,37 +91,80 @@ object DagOperations {
     })
   }
 
+  private val blockMetadataByNumDecreasing: Ordering[BlockMetadata] =
+    (l: BlockMetadata, r: BlockMetadata) => {
+      def compareByteString(l: ByteString, r: ByteString): Int =
+        l.hashCode().compareTo(r.hashCode())
+
+      val ln = l.blockNum
+      val rn = r.blockNum
+      rn.compare(ln) match {
+        case 0 => compareByteString(l.blockHash, r.blockHash)
+        case v => v
+      }
+    }
+
+  private def getParents[F[_]: Monad](
+      dag: BlockDagRepresentation[F],
+      p: BlockMetadata
+  ): F[Set[BlockMetadata]] =
+    p.parents.traverse(dag.lookup).map(_.toSet.flatten)
+
   /**
-    * Conceptually, the LCA is the lowest point at which the histories of b1 and b2 diverge.
-    * We compute by finding the first block that is the "lowest" (has highest blocknum) common block.
+    * LCA is the "lowest" (has highest blocknum) block which is an ancestor to both blocks.
     */
   def lowestCommonAncestorF[F[_]: Monad](
       b1: BlockMetadata,
       b2: BlockMetadata,
       dag: BlockDagRepresentation[F]
   ): F[BlockMetadata] = {
 
-    implicit val blockMetadataByNumDecreasing: Ordering[BlockMetadata] =
-      (l: BlockMetadata, r: BlockMetadata) => {
-        def compareByteString(l: ByteString, r: ByteString): Int =
-          l.hashCode().compareTo(r.hashCode())
+    implicit val ordering = blockMetadataByNumDecreasing
 
-        val ln = l.blockNum
-        val rn = r.blockNum
-        rn.compare(ln) match {
-          case 0 => compareByteString(l.blockHash, r.blockHash)
-          case v => v
-        }
+    type LeftRightAncestors = (SortedSet[BlockMetadata], SortedSet[BlockMetadata])
+
+    def extractParentsFromHighestNumBlock(blocks: LeftRightAncestors): F[LeftRightAncestors] = {
+      val (leftAncestors, rightAncestors) = blocks
+      val (leftMin, leftTail)             = (leftAncestors.head, leftAncestors.tail)
+      val (rightMin, rightTail)           = (rightAncestors.head, rightAncestors.tail)
+      if (ordering.compare(leftMin, rightMin) <= 0) {
+        getParents(dag, leftMin).map(newBlocks => (leftTail ++ newBlocks, rightAncestors))
+      } else {
+        getParents(dag, rightMin).map(newBlocks => (leftAncestors, rightTail ++ newBlocks))
       }
+    }
+
+    if (b1 == b2) {
+      b1.pure[F]
+    } else {
+      val start = (SortedSet(b1), SortedSet(b2))
+      Monad[F]
+        .iterateWhileM(start)(extractParentsFromHighestNumBlock) {
+          case (leftAncestors, rightAncestors) =>
+            leftAncestors.head != rightAncestors.head
+        }
+        .map(_._1.head)
+    }
+  }
+
+  /**
+    * Conceptually, the LCfPA is the lowest point at which the histories of b1 and b2 diverge.
+    * We compute by finding the first block that is the "lowest" (has highest blocknum) block common
+    * for both blocks' parents.
+    */
+  def lowestCommonForParentsAncestorF[F[_]: Monad](
```

Comment:
I think `for parents` is not precise here, as for the following:

```
A   B
 \ / \
  C   D
   \ / \ 
    E   F
     \ /
      G
```
LCFPA(A, B) == G, while - if we take parents of A and B, their LCA(A, B) == E

I think this should be called 'lowest universal common ancestor', which reflects the fact we require the ancestor to be the "Adam" for A and B, or "the nearest common ancestor of A, B, and all their ancestors"

Reply:
Renamed it to lowestUniversalCommonAncestor and clarified the comment a bit further